### ISSUE TITLE
Implement of ResNet

### ISSUE THREAD

**Issue by PengBY91** [sseung0703/Knowledge_distillation_methods_wtih_Tensorflow#5]
It seems that the implementation of ResNet in your project is different from the official implementation [https://github.com/tensorflow/models/blob/master/research/resnet/resnet_model.py](url). Your implementation of ResNet have much more parameters than official implementation.

**Comment by sseung0703** [comment]
I already mentioned it in network architecture. they have twice depth than the original one.

**Comment by cupwater** [comment]
Did you try original ResNet for SVD or any other KD-based methods?  I have just run your code, and I get 72.9% top1 valid accuracy when trained with cross-entropy loss, and 75.5% top1 valid when trained with soft_logits for the student (ResNet-8). When trained with KD-EID, I get 75.1% top valid accuracy. The only difference is that I changed the Temperature to 5 and remove  classifier_loss both for Soft_logits and KD-EID. The teacher model is ResNet32 and I just use the params ResNet32.mat provided in code. The results is much different from those the project reports. Is there any point I have not realized? Hope for your positive response. Sincerely, thanks!

**Comment by sseung0703** [comment]
For Soft-logits I didn't find eagerly optimal hyper-parameters so I think you can find a better configuration. And I don't know exactly how you modified the codes, but it is not possible to train KD-EID without cross entropy loss which named "class_loss" and If student and teacher feature map's depths are different, knowledge transfer is not possible. so I think you had some mistakes.

**Comment by PengBY91** [comment]
sincerely thanks，I will check my code carefully.

**Comment by cupwater** [comment]
Thanks! I have implemented the original ResNet 

> For Soft-logits I didn't find eagerly optimal hyper-parameters so I think you can find a better configuration. And I don't know exactly how you modified the codes, but it is not possible to train KD-EID without cross entropy loss which named "class_loss" and If student and teacher feature map's depths are different, knowledge transfer is not possible. so I think you had some mistakes.

Sorry, I have made a mistake that I take the KD-SVD as KD-EID. The results of KD-EID should be corrected as KD-SVD. The modification as follows: batch_size=64, learning_rate_scheduler=[0.4, 0.7, 0.9], weight_decay=1e-4, T=5, class_loss=0.

**Comment by sseung0703** [comment]
I don't understand what you say. KD-EID and KD-SVD are almost the same algorithms and they require cross entropy loss and the same depth of the feature map.
But you said that you modify the student network to have half size depth but not modify the teacher network.
Soft-logits is fine cause it only requires the same logits, but KD-EID and KD-SVD are not.

**Comment by PengBY91** [comment]
Thanks. KD-SVD is an excellent work, and I plan to follow it. What I want to is to figure out the details of KD-SVD. Sincerely thanks for your positive response.

**Comment by PengBY91** [comment]
> I don't understand what you say. KD-EID and KD-SVD are almost the same algorithms and they require cross entropy loss and the same depth of the feature map.
> But you said that you modify the student network to have half size depth but not modify the teacher network.
> Soft-logits is fine cause it only requires the same logits, but KD-EID and KD-SVD are not.

Hi, I have carefully checked my code, and run again KD-EID and KD-SVD **without class_loss**, it really can run and I get 74.56% (best) for KD-SVD and 74.35% (best) for KD-EID. Both significantly outperform student trained by class loss. BTW, I **have not changed** the student or teacher network. When **combining with Soft_logits**, KD-SVD or KD-EID can achieve higher performance. Another interesting phenomenon is that when combined with class loss, soft_logit or KD-SVD performs worse, which is worth more experiments and analysis. Nice project, it is of much help for me.


the code of KD-SVD or KD-EID as follows :


            total_loss = reg_loss + distillation_loss
            tf.summary.scalar('loss/total_loss', total_loss)
            tf.summary.scalar('loss/distillation_loss', distillation_loss)
            gradients  = optimize.compute_gradients(class_loss,  var_list = variables)
            gradient_wdecay = optimize.compute_gradients(reg_loss,  var_list = variables)
            gradient_dist   = optimize.compute_gradients(distillation_loss, var_list = variables)
            with tf.variable_scope('clip_grad'):
                for i, gc, gw, gd in zip(range(len(gradients)),gradients,gradient_wdecay,gradient_dist):
                    gw = 0. if gw[0] is None else gw[0]
                    if gd[0] != None:
                        norm = tf.sqrt(tf.reduce_sum(tf.square(gc[0])))*sigmoid(epoch, 0)
                        gradients[i] = (gc[0] + gw + tf.clip_by_norm(gd[0], norm), gc[1])
                    else:
                        gradients[i] = (gc[0] + gw, gc[1])

**Comment by sseung0703** [comment]
You still use class loss. As you can see the gradients propagated by class loss is used. the only difference is the summation of loss. How many times did you train the model? you should at least three times to reduce variance.

**Comment by PengBY91** [comment]
👌, get it. I have run it only 1 time. 
btw, does kd-svd require the same channels for teacher and student？because there are some errors when i reduce the channels of student.

**Comment by sseung0703** [comment]
Yes. as I mentioned above.
How about reading my paper? There are all of the details.

**Comment by PengBY91** [comment]
> Yes. as I mentioned above.
> How about reading my paper? There are all of the details.

Ok, it is worth to read and follow.

**Comment by HobbitLong** [comment]
> I already mentioned it in network architecture. they have twice depth than the original one.

Hi, @sseung0703 ,

Nice work!

Just want to be precise, I thought your model is 4 times wider than the original resnet, rather than 2 times? The original resnet has 16, 32 and 64 channels for each of the residual block, while here your model has 64, 128 and 256 instead. Please correct me if I am wrong.
